ros2_batch_job/__main__.py

```diff
@@ -356,6 +366,104 @@ def build_and_test(args, job):
     # return 0 if ret_test == 0 and ret_testr == 0 else 1
     return 0
 
+'''
+Add sanitizer based flags to colcon build for appropriate packages.
+'''
+def update_build_cmd_based_on_sanitizer_args(cmd, args):
+    sanitizer_flags = '-fsanitize={} -pthread'.format(args.enable_sanitizer_type)
+    additional_compiler_flags = '-g -fno-omit-frame-pointer'
+    sanitizer_cmake_args = [
+    '-DCMAKE_CXX_FLAGS="{} {}"'.format(sanitizer_flags, additional_compiler_flags),
+    '-DCMAKE_C_FLAGS="{} {}"'.format(sanitizer_flags, additional_compiler_flags),
+    '-DCMAKE_EXE_LINKER_FLAGS="{}"'.format(sanitizer_flags),
+    '-DCMAKE_MODULE_LINKER_FLAGS="{}"'.format(sanitizer_flags),
+    '-DCMAKE_SHARED_LINKER_FLAGS="{}"'.format(sanitizer_flags)
+    ]
+
+    '''
+    Build command is split into build upto regex packages, build regex packages and then build
+    packages above regex packages. This ensures that we do not rebuild all the packages again with sanitizer flags.
+    '''
```

Comment:
This approach is not going to work for an arbitrary regex which could select packages throughout the dependency graph.

Reply:
Can you give an example here? I am sorry I am new to to colcon build concepts :(.

Reply:
If you have three packages with a dependency chain:

`A -> B -> C`

And the regex matches A and C.

Reply:
Oh got it, thanks. We may be alright to continue with this logic as we ultimately want to test sanitization on just A and C and hence do not need B to be sanitized. Also, in the previous approach(build all without sanitizer flags and build just regex packages with sanitizer flags), we ended up in a similar state where A and C are sanitized but B is not. Let me know if I am missing something here.

It is definitely true that there may be some occasions in which both the approaches might break. For example, if colcon test on sanitized A needs B to be sanitized as well.

Reply:
The approach will still build some of the packages twice. Also hard coding all this extra logic in this repo doesn't look appealing to me.

Reply:
> Taking a step back I would prefer an approach where a user can perform the same build locally - without having to duplicate all this logic / settings. So e.g. all the necessary CMake arguments could be stored in a colcon mixin file (as it is already the case for some: https://github.com/colcon/colcon-mixin-repository/blob/3e627e0fa30db85aea05a50e2c61a9832664d236/asan.mixin - not sure if the differences are necessary and should be added to the existing mixin?). Based on that there should be a way to specify that the mixin should be applied to a set of packages (needs to be implenmnted) and after that a single build pass should be all it takes.
> 
> This would make it easy for users to run the very same locally as well as reduce the custom logic necessary to integration it into the Jenkins jobs.

Makes sense. This way all the packages will be built only once. I am assuming this would be a change in colcon itself (adding an option to add a mixin for a set of packages). Do you think we should handle it as a separate issue?

Reply:
> The approach will still build some of the packages twice. Also hard coding all this extra logic in this repo doesn't look appealing to me.

One naive way to avoid the rebuild would be something this:
Step 1:   Build all pkgs upto regex and skip regex pkgs
Step 2:   Build regex pkgs
Step 3: Build pkgs above regex pkgs

Definitely ugly but will avoid the rebuild

Reply:
> I am assuming this would be a change in colcon itself (adding an option to add a mixin for a set of packages). 

Yes, that would require support for that use case in `colcon`. I am just not sure how exactly it would look like. The complexity for controlling this using command line options is too high imo. I will need to think about how this fits in with the already existing feature like metadata, mixins, defaults as well as other feature ideas like profiles...

> One naive way to avoid the rebuild would be something this:

That approach doesn't work which I tried to point out above. If your regex matched packages have non-matches packages in between 3 steps are not sufficient. In a trivial case you would need 5 and in a more complex case up to N separate steps.

Reply:
What is the downside of not sanitizing the packages in between two regex matched packages? Instrumentation will not be performed on these packages and hence normal malloc() and other functions will be called. Are we targetting the use case where all regex matched packages and the packages between them are sanitized?

Reply:
> Step 1: Build all pkgs upto regex and skip regex pkgs
> Step 2: Build regex pkgs
> Step 3: Build pkgs above regex pkgs

The problem is not that some packages are not sanitized. The build in the first step will just fail. In my previously mentioned example (A -> B -> C where the regex matches A and C) step 1 tries to build A and C. But since B hasn't been built yet but is a dependency of A the build will fail.

Reply:
Sorry for another rookie question, but I assumed that B will be built before A is built. Is it not what "--packages-up-to" is supposed to do? 
  --packages-up-to [PKG_NAME [PKG_NAME ...]]
                        Only process a subset of packages and their recursive
                        dependencies

Reply:
Your step explicitly states: `Build all pkgs upto regex and skip regex pkgs`.

So when you select "up to" A and C you will select all three packages and applying the "skip regex pkgs" you will only have package B left. But that package can't be built without its dependency C being built first.

Reply:
Ahh.. thanks a lot for pointing it out. Makes sense, so skipping is not an option. Building the regex packages twice is safer.